modules/tinymce/src/core/test/ts/browser/EditorUploadTest.ts

```diff
@@ -358,8 +358,8 @@ describe('browser.tinymce.core.EditorUploadTest', () => {
       callCount++;
 
       if (callCount === 2) {
-        // Note: This is 2 as the removal of the image also triggers the addition of an undo level and a change event
-        assertEventsLength(2);
+        // Note: This is 1 as only the removal of the image triggers the addition of an undo level and a change event
```

Comment:
This is probably a side-effect of you changing the placement of the `change` call and proves that you've changed behaviour which isn't what we want here.

Reply:
I did a fix that seems to preserve the old behavior but I change it back with this [commit](https://github.com/tinymce/tinymce/pull/7840/commits/df5c497c1985ac4f54aaefc5c2d4bb3c5ff333e9) because of this [comment](https://github.com/tinymce/tinymce/pull/7840#discussion_r894208749).

At the moment the only way to preserve the behavior that I found was my [previous fix](https://github.com/tinymce/tinymce/pull/7840/commits/df5c497c1985ac4f54aaefc5c2d4bb3c5ff333e9#diff-b611e1623b7290a18e6eac5c3465926dae2ac0f148142325b5e2f08ca8636e39L156-L158), I made this [commit](https://github.com/tinymce/tinymce/pull/7840/commits/8e65e8e9725d6edb4284711937ecd20a079b3996) to fix it better than how was in the previous code, but I'd also like the opinion of @spocke since he wrote this https://github.com/tinymce/tinymce/pull/7840#discussion_r894208749

Reply:
@lnewson I really think the behavior we have here is wrong that if you upload the same image in parallel and remove the image you get two change events. I would expect to only get one. I doubt anyone if depending on us producing two change events in a row for that particular case.
Switching this code from the current compare with undo levels approach would make that 3 times even more incorrect. So the code needs to be tweaked if we want it to use the new dispatchChange event.
The transact call could be enough to detect the remove image change and fire the change event when it's needed when a new undo level is added.
For uploading images that doesn't get removed then there should be more no new undo level but a change event should be fired so that people can save after an image gets uploaded. At least that is my understanding of this code and problem with changing it.

Reply:
@spocke per the comment, the first change is to notify that the existing undomanager state has changed and the second is related to the additional remove steps which is really a separate process and hence the new undomanager state. So there definitely should be the 2 change events here, as people might need to handle that state change.

Reply:
Also the original logic that was here previously would still have resulted in 2 change events firing because it was this:
```
        if (imagesToRemove.length > 0 && !Rtc.isRtc(editor)) {
          editor.undoManager.transact(() => {
            Arr.each(imagesToRemove, (element) => {
              editor.dom.remove(element);
              blobCache.removeByUri(element.src);
            });
          });
        }

        if (shouldDispatchChange) {
          editor.undoManager.dispatchChange();
        }
```
That meant that you'd then get 2 change events for the exact same undo level which definitely wasn't what we wanted.

Reply:
Yes, and that is the thing I think is wrong. If your upload handler removes an image then removes the same image again only one change should be fired. The scanner will find one image, then do an async upload, then remove the image when the first one resolves then tries to remove it again when the second one gets resolved. So shouldDispatchChange could check both if there are no url mutations or node remove mutations then don't fire a change event. It's more that we need to tweak the code to not produce 3 change events then we might as well tweak it to not produce 2. It can't be uploaded and removed at the same time since the upload either passes or fails if it passes then it's an url mutation to change the url to the new one if it fails and has the remove flag then it should remove the image and add an extra undo level since it's a visual change. It would still fire 2 events if you for example upload one image that passed and one that fails and got removed. But I guess we can go all though hoops to get this to work exactly as it did before dispatching additional unnecessary change events if we need to but then I think we might not change this code at all to use the new dispatchChangeEvent since we are afraid of potential behavioral changes and since it's not comparing the undo stack as before it would be a behavioral change in theory if someone has some weird dependency of it having to do that.

Reply:
That's a fair point about `uploadImages` being a single operation that either fails or succeeds. I'm okay with making it only fire the single change here (though as I said the previous logic didn't do that) and the very small risk of breaking that one edge case as long as we put a changelog entry in about it as well.

Reply:
At the moment the new code behave as the old one, should I left this behavior or should I remove the check that trigger the change and add a change to the changelog that says:
```
- Now on `uploadImages` does not trigger `change` event if there is an error on upload #TINY-8641
```
?

Reply:
Sorry I checked again the behaviour, and the 2 events are triggered for the same reason so the correct entry probably would be:
```
- Now on `uploadImages` does not trigger `change` event 2 times if there is an error on upload #TINY-8641
```

Reply:
I believe adding a changelog fix entry is the right way forward. It's become an intentional change, and we need some indication to integrators that they have something watch out for. What you wrote is very close, but I suggest:

```
<under the 'fixed' heading>
- `uploadImages` no longer triggers two `change` events if there is an error on upload #TINY-8641
- ```

Reply:
I added the new entry too